File: Ejercicio_N1/Piramide.cpp
/**
 * @file    Piramide.cpp
 * @brief   Expansión Binomial y Evaluación de Polinomio
 *
 * Este programa genera los coeficientes del binomio (1 + x)^n mediante
 * la fórmula de combinaciones C(n, k), construye la representación
 * textual del polinomio resultante, evalúa f(x) paso a paso para un
 * valor dado de x, y mide los tiempos de ejecución de cada etapa.
 *
 * Los resultados de tiempo se escriben adicionalmente en el archivo
 * 'resultado.txt' para comparación con la versión en Python.
 *
 * @author  Jesús Báez
 * @date    2026
 */

#include <chrono>
#include <cmath>
#include <fstream>
#include <iomanip>
#include <iostream>

using namespace std;
using namespace std::chrono;

// ===========================================================================
//  Prototipos de funciones
// ===========================================================================

double *generarCoeficientes(int n);
void mostrarPolinomio(const double *coeficientes, int n);
double calcularFx(const double *coeficientes, int n, double x);
void guardarResultados(int n, double x, const double *coeficientes,
                       double resultadoFx, double tiempoGeneracion,
                       double tiempoTotal);

// ===========================================================================
//  Implementación de funciones
// ===========================================================================

/**
 * @brief Genera los coeficientes binomiales C(n, 0) a C(n, n).
 *
 * Utiliza la relación recursiva:
 *     C(n, k) = C(n, k-1) * (n - k + 1) / k
 *
 * @param  n  Exponente del binomio (entero no negativo).
 * @return    Arreglo dinámico de tamaño n+1 con los coeficientes.
 *
 * @note El llamador es responsable de liberar la memoria con delete[].
 */
double *generarCoeficientes(int n) {
  double *coeficientes = new double[n + 1];
  coeficientes[0] = 1; // C(n, 0) = 1

  for (int k = 1; k <= n; k++) {
    coeficientes[k] = coeficientes[k - 1] * (n - k + 1) / k;
  }

  return coeficientes;
}

/**
 * @brief Muestra la representación textual del polinomio en consola.
 *
 * El formato generado es:
 *     C(n,0)*x^0 + C(n,1)*x^1 + ... + C(n,n)*x^n
 *
 * @param coeficientes  Arreglo de coeficientes binomiales.
 * @param n             Grado del polinomio.
 */
void mostrarPolinomio(const double *coeficientes, int n) {
  cout << "\nPolinomio resultante:" << endl;

  for (int i = 0; i <= n; i++) {
    if (i == 0) {
      cout << coeficientes[i] << "*x^" << i;
    } else {
      cout << " + " << coeficientes[i] << "*x^" << i;
    }
  }

  cout << endl;
}

/**
 * @brief Evalúa el polinomio f(x) = Σ C(n, k) * x^k mostrando cada paso.
 *
 * @param coeficientes  Arreglo de coeficientes binomiales.
 * @param n             Grado del polinomio.
 * @param x             Valor en el que se evalúa el polinomio.
 * @return              Resultado numérico de f(x).
 */
double calcularFx(const double *coeficientes, int n, double x) {
  cout << "\nCalculo paso a paso de f(x):" << endl;

  double resultado = 0;

  for (int i = 0; i <= n; i++) {
    double termino = coeficientes[i] * pow(x, i);
    resultado += termino;

    cout << fixed << setprecision(0);
    cout << "Paso " << i << ": " << coeficientes[i] << "*(" << x << "^" << i
         << ") = " << termino << endl;
  }

  return resultado;
}

/**
 * @brief Escribe todos los resultados al archivo 'resultado.txt'.
 *
 * @param n                Exponente del binomio.
 * @param x                Valor en el que se evaluó el polinomio.
 * @param coeficientes     Arreglo de coeficientes binomiales.
 * @param resultadoFx      Resultado numérico de f(x).
 * @param tiempoGeneracion Tiempo de generación de coeficientes en segundos.
 * @param tiempoTotal      Tiempo total de ejecución en segundos.
 */
void guardarResultados(int n, double x, const double *coeficientes,
                       double resultadoFx, double tiempoGeneracion,
                       double tiempoTotal) {
  ofstream archivo("resultado.txt", ios::app);

  archivo << "==================================================" << endl;
  archivo << "--- Resultados C++ ---" << endl;
  archivo << "==================================================" << endl
          << endl;

  archivo << fixed << setprecision(0);
  archivo << "n = " << n << endl;
  archivo << "x = " << x << endl << endl;

  // Polinomio
  archivo << "Polinomio resultante:" << endl;
  for (int i = 0; i <= n; i++) {
    if (i == 0) {
      archivo << coeficientes[i] << "*x^" << i;
    } else {
      archivo << " + " << coeficientes[i] << "*x^" << i;
    }
  }
  archivo << endl << endl;

  // Calculo paso a paso
  archivo << "Calculo paso a paso de f(x):" << endl;
  for (int i = 0; i <= n; i++) {
    double termino = coeficientes[i] * pow(x, i);
    archivo << "Paso " << i << ": " << coeficientes[i] << "*(" << x << "^" << i
            << ") = " << termino << endl;
  }

  // Resultado final
  archivo << endl;
  archivo << scientific << setprecision(15);
  archivo << "Resultado final: f(" << fixed << setprecision(0) << x
          << ") = " << scientific << setprecision(15) << resultadoFx << endl
          << endl;

  // Tiempos
  archivo << fixed << setprecision(6);
  archivo << "Tiempo generacion coeficientes: " << tiempoGeneracion
          << " segundos" << endl;
  archivo << "Tiempo total calculo + evaluacion: " << tiempoTotal << " segundos"
          << endl;

  archivo.close();

  cout << "\nResultados guardados en 'resultado.txt'" << endl;
}

// ===========================================================================
//  Programa principal
// ===========================================================================

/**
 * @brief Punto de entrada del programa.
 */
int main() {
  int n;
  double x;

  // ---- Lectura de datos de entrada ------------------------------------
  cout << "Ingrese un numero entero no negativo n: ";
  cin >> n;

  // ---- Generación de coeficientes (con medición de tiempo) ------------
  auto inicioGeneracion = high_resolution_clock::now();
  double *coeficientes = generarCoeficientes(n);
  auto finGeneracion = high_resolution_clock::now();

  double tiempoGeneracion =
      duration_cast<duration<double>>(finGeneracion - inicioGeneracion).count();

  // ---- Lectura del valor de x ----------------------------------------
  cout << "Ingrese el valor de x: ";
  cin >> x;

  // ---- Mostrar el polinomio resultante --------------------------------
  mostrarPolinomio(coeficientes, n);

  // ---- Evaluación de f(x) (con medición de tiempo) --------------------
  auto inicioEval = high_resolution_clock::now();
  double resultado = calcularFx(coeficientes, n, x);
  auto finEval = high_resolution_clock::now();

  double tiempoTotal =
      duration_cast<duration<double>>(finEval - inicioEval).count();

  // ---- Mostrar resultados en consola ----------------------------------
  cout << fixed << setprecision(15);
  cout << "\nf(" << x << ") = " << resultado << endl;
  cout << "\nTiempo generaci  on coeficientes: " << tiempoGeneracion
       << " segundos";
  cout << "\nTiempo total calculo + evaluacion: " << tiempoTotal << " segundos"
       << endl;

  // ---- Guardar resumen en archivo -------------------------------------
  guardarResultados(n, x, coeficientes, resultado, tiempoGeneracion,
                    tiempoTotal);

  // ---- Liberar memoria dinámica ---------------------------------------
  delete[] coeficientes;

  return 0;
}

File: Ejercicio_N1/Piramide.py
"""
Piramide.py - Expansión Binomial y Evaluación de Polinomio

Este módulo genera los coeficientes del binomio (1 + x)^n mediante la
fórmula de combinaciones C(n, k), construye la representación textual
del polinomio resultante, evalúa f(x) paso a paso para un valor dado
de x, y mide los tiempos de ejecución de cada etapa.

Los resultados de tiempo se escriben adicionalmente en el archivo
'resultado.txt' para comparación con la versión en C++.

Autor:  Jesús Báez
Fecha:  2026
"""

import time


# ===========================================================================
#  Funciones auxiliares
# ===========================================================================

def generar_coeficientes(n):
    """
    Genera la lista de coeficientes binomiales C(n, 0), C(n, 1), ..., C(n, n).

    Utiliza la relacion recursiva:
        C(n, k) = C(n, k-1) * (n - k + 1) / k

    Parámetros
    ----------
    n : int
        Exponente del binomio (entero no negativo).

    Retorna
    -------
    list[int]
        Lista con n+1 coeficientes binomiales.
    """
    coeficientes = [1]  # C(n, 0) = 1

    for k in range(1, n + 1):
        siguiente = coeficientes[k - 1] * (n - k + 1) // k
        coeficientes.append(siguiente)

    return coeficientes


def mostrar_polinomio(coeficientes):
    """
    Construye la representacion textual del polinomio.

    El formato generado es:
        C(n,0)*x^0 + C(n,1)*x^1 + ... + C(n,n)*x^n

    Parámetros
    ----------
    coeficientes : list[int]
        Lista de coeficientes binomiales.

    Retorna
    -------
    str
        Cadena con la representacion del polinomio.
    """
    terminos = []

    for exponente, coef in enumerate(coeficientes):
        if exponente == 0:
            terminos.append(f"{coef}*x^{exponente}")
        else:
            terminos.append(f" + {coef}*x^{exponente}")

    return "".join(terminos)


def calcular_fx(coeficientes, x):
    """
    Evalúa el polinomio f(x) = Σ C(n, k) * x^k  mostrando cada paso.

    Parámetros
    ----------
    coeficientes : list[int]
        Lista de coeficientes binomiales.
    x : int
        Valor en el que se evalúa el polinomio.

    Retorna
    -------
    int | float
        Resultado numérico de f(x).
    """
    print("\nCalculo paso a paso de f(x):")

    resultado = 0

    for exponente, coef in enumerate(coeficientes):
        termino = coef * (x ** exponente)
        resultado += termino
        print(f"Paso {exponente}: {coef}*({x}^{exponente}) = {termino}")

    return resultado


def guardar_resultados(n, x, coeficientes, resultado_fx,
                       tiempo_generacion, tiempo_total):
    """
    Escribe todos los resultados al archivo 'resultado.txt'.

    Parámetros
    ----------
    n : int
        Exponente del binomio.
    x : int
        Valor en el que se evaluó el polinomio.
    coeficientes : list[int]
        Lista de coeficientes binomiales.
    resultado_fx : int | float
        Resultado numérico de f(x).
    tiempo_generacion : float
        Tiempo de generación de coeficientes en segundos.
    tiempo_total : float
        Tiempo total de ejecucion en segundos.
    """
    with open("resultado.txt", "a", encoding="utf-8") as archivo:
        archivo.write("=" * 50 + "\n")
        archivo.write("--- Resultados Python ---\n")
        archivo.write("=" * 50 + "\n\n")

        archivo.write(f"n = {n}\n")
        archivo.write(f"x = {x}\n\n")

        # Polinomio
        archivo.write("Polinomio resultante:\n")
        archivo.write(mostrar_polinomio(coeficientes) + "\n\n")

        # Cálculo paso a paso
        archivo.write("Calculo paso a paso de f(x):\n")
        for exponente, coef in enumerate(coeficientes):
            termino = coef * (x ** exponente)
            archivo.write(
                f"Paso {exponente}: {coef}*({x}^{exponente}) = {termino}\n"
            )

        # Resultado final
        archivo.write(f"\nResultado final: f({x}) = {resultado_fx:.15e}\n\n")

        # Tiempos
        archivo.write(
            f"Tiempo generacion coeficientes: {tiempo_generacion:.6f} segundos\n"
        )
        archivo.write(
            f"Tiempo total calculo + evaluacion: {tiempo_total:.6f} segundos\n"
        )

    print("\nResultados guardados en 'resultado.txt'")


# ===========================================================================
#  Programa principal
# ===========================================================================

def main():
    """Punto de entrada del programa."""

    # ---- Lectura de datos de entrada ------------------------------------
    n = int(input("Ingrese un numero entero no negativo n: "))
    x = int(input("Ingrese el valor de x: "))

    # ---- Generacion de coeficientes (con medicion de tiempo) ------------
    inicio_generacion = time.time()
    coeficientes = generar_coeficientes(n)
    fin_generacion = time.time()

    tiempo_generacion = fin_generacion - inicio_generacion

    # ---- Mostrar el polinomio resultante --------------------------------
    print("\nPolinomio resultante:")
    print(mostrar_polinomio(coeficientes))

    # ---- Evaluacion de f(x) (con medicion de tiempo total) --------------
    resultado_fx = calcular_fx(coeficientes, x)
    fin_total = time.time()

    tiempo_total = fin_total - inicio_generacion  # Desde el inicio de generacion

    # ---- Mostrar resultados en consola ----------------------------------
    print(f"\nResultado final: f({x}) = {resultado_fx:.15e}")
    print(f"\nTiempo generacion coeficientes: {tiempo_generacion:.6f} segundos")
    print(f"Tiempo total calculo + evaluacion: {tiempo_total:.6f} segundos")

    # ---- Guardar resumen en archivo -------------------------------------
    guardar_resultados(n, x, coeficientes, resultado_fx,
                       tiempo_generacion, tiempo_total)


if __name__ == "__main__":
    main()
